Integrate QUIC text from DPLPMTUD (#3693)
quicwg/base-drafts: draft-ietf-quic-transport.md

The QUIC packet size includes the QUIC header and protected payload, but not the
UDP or IP header.

> QUIC defines the minimum QUIC packet size as at least 1280 bytes.

Reviewer: ```suggestion
QUIC depends upon a minimum packet size as at least 1280 bytes.
```

Author: Fine, except shouldn't this be: /as at/of at/

This is the IPv6 minimum size
{{?RFC8200}} and is also supported by most modern IPv4 networks.
Assuming the minimum IP header size, this results in a

> QUIC maximum packet size of 1232 bytes for IPv6 and 1252 bytes for IPv4. 

Author: I like clarity about constants and variables, the text in DPLPMTUD says: "The Maximum Packet Size (MPS) is the largest size of application data block that can be sent across a network path by a PL using a single Datagram.", so for QUIC, maybe this could read: 
"The QUIC maximum packet size is the largest size of QUIC packet that can be sent across a network path using a single packet. This is discovered using PMTUD/DPLPMTUD"?

> QUIC maximum packet size of 1232 bytes for IPv6 and 1252 bytes for IPv4. 

Reviewer 2: I could live with that.

A client MUST expand the payload of all UDP datagrams carrying Initial packets

> to at least the minimum QUIC packet size, by adding PADDING frames to the Initial packet or by

Reviewer: Keep this as 1200 bytes.

```suggestion
to at least 1200 bytes, by adding PADDING frames to the Initial packet or by
```

over time.  The largest UDP payload an endpoint sends at any given time is
referred to as the endpoint's maximum packet size.

An endpoint SHOULD use Datagram Packetization Layer PMTU Discovery
({{!DPLPMTUD=I-D.ietf-tsvwg-datagram-plpmtud}}) or implement Path MTU Discovery
(PMTUD) {{!RFC1191}} {{!RFC8201}} to determine whether the path to a destination
will support a desired message size without fragmentation.
In the absence of these mechanisms, QUIC endpoints SHOULD NOT send IP packets
larger than the minimum QUIC packet size. 

All QUIC
packets (except for PMTUD/DPLPMTUD probe packets) SHOULD be sized to fit within the
maximum packet size to avoid the packet being fragmented or dropped
{{?RFC8085}}.

If a QUIC endpoint determines that the PMTU between any pair of local and remote

> IP addresses has fallen below the minimum QUIC packet size, it MUST immediately 

Reviewer: Please keep the existing wording of "smallest allowed maximum packet size", since that's very different from the minimum QUIC packet size.

```suggestion
IP addresses has fallen below the smallest allowed maximum packet size, it MUST immediately 
```

### Handling of ICMP PTB Messages by PMTUD {#icmp-pmtud}

PMTUD {{!RFC1191}} {{!RFC8201}} relies on reception of ICMP messages
(e.g., IPv6 Packet Too Big messages) that indicate when a packet is dropped
because it is larger than the local router MTU. DPLPMTUD can also optionally use
these messages.  This use of ICMP messages is potentially vulnerable to off-path
attacks that successfully guess the addresses used on the path and reduce the
PMTU to a bandwidth-inefficient value.

An endpoint MUST ignore an ICMP message that claims the PMTU has decreased below
the minimum QUIC packet size bytes.

The requirements for generating ICMP ({{?RFC1812}}, {{?RFC4443}}) state that the
quoted packet should contain as much of the original packet as possible without
exceeding the minimum MTU for the IP version.  The size of the quoted packet can
actually be smaller, or the information unintelligible, as described in Section
1.1 of {{!DPLPMTUD}}.

QUIC endpoints using PMTUD SHOULD validate ICMP messages to protect from off-path injection
as specified in {{!RFC8201}} and Section 5.2 of {{!RFC8085}}. This validation
SHOULD use the quoted packet supplied in the payload of an ICMP message to

> associate the message with a corresponding transport connection (e.g., {{!DPLPMTUD}}).

Author: I'd prefer "see Section 4.6.1 of {{!DPLPMTUD}}."

> associate the message with a corresponding transport connection (e.g., {{!DPLPMTUD}}).

Reviewer 2: I'm happy with either, as I said.

ICMP message validation MUST include matching IP addresses and UDP ports
{{!RFC8085}} and, when possible, connection IDs to an active QUIC session.
The endpoint SHOULD ignore all ICMP messages that fail validation.

An endpoint MUST NOT increase PMTU based on ICMP messages; see Section 3, clause
6 of {{!DPLPMTUD}}.  Any reduction in the QUIC maximum packet size in response
to ICMP messages MAY be provisional until QUIC's loss detection algorithm
determines that the quoted packet has actually been lost.

> ### PMTUD Probes with Handshake packets

Reviewer 2: Please move this back.  It belongs where it was taken from.

I realize that you now have a section for PMTUD parallel to the section for PLPMTUD, and this allows you to make the structure cleaner (with two subsections and not one, which I agree is nice), but this text is relevant to both.

(I think that this move points out that the specific use of Handshake isn't the ideal. Though Initial packets might be a bad idea, using 0-RTT is also fine.)

Author: So, the idea of sending a larger probe before the connection is established is something that probably needs more discussion for both cases.

Reviewer 2: This is not about when, but how.  You can send a bogus Handshake packet after the connection is up.

Author: OK - Can we say that explicitly?

Reviewer 2: Please take a look at #3702 for a proposed tweak.

first part of the datagram will be quoted in that message.  If the source
connection ID is within the quoted portion of the UDP datagram, that could be
used for routing.

## Datagram Packetization Layer PMTU Discovery

When implementing the algorithm in Section 5 of {{!DPLPMTUD}}, the
initial value of BASE_PMTU SHOULD be consistent with the minimum QUIC
packet size. The MIN_PLPMTU is the same as the BASE_PMTU.

QUIC endpoints implementing DPLPMTUD maintain a maximum packet size 
(DPLPMTUD MPS) for each combination of local and remote IP
addresses. 
If a QUIC endpoint determines that the PLPMTU between any pair of local
and remote IP addresses has fallen below the size needed to support
the minimum QUIC packet size (BASE_PLPMTU), it MUST immediately cease
sending QUIC packets, except for DPLPMTUD probe packets, on the affected
path. An endpoint MAY terminate the connection if an alternative
path cannot be found.

###  DPLPMTUD and Initial Connectivity

From the perspective of DPLPMTUD, QUIC transport is an 
acknowledged PL. A sender can therefore enter the DPLPMTUD BASE
state when the QUIC connection handshake has been completed and

> the endpoint has established a 1-RTT key.

Author: This text was proposed by Magnus in a tsvwg thread. We should discuss this.

> the endpoint has established a 1-RTT key.

Reviewer 3: So, this was an attempt from me to correct something that looked even more broken in the earlier text in the TSVWG document. If handshake is complete is a good term lets use it. 

So I think it is relevant to at least go through the implication of sending DPLPMTUD probes prior to handshake complete. So one possible implementation here could be to first send your initial packet with 1280 as MPS. Then you use the rest of your IW=10 to send a broad side of Path MTU probes of various sizes checking a number of common sizes up to the client's local interface's MPS. So in relation to recovery that defines the initial window to 10*max_datagram_size with a limit of max (14720, 2*max_datagram_size). I make an assumption that one will in this case start with max_datagram_size=1200. Thus one can fit some number of probes. So allow DPLPMTUD to use 0-RTT keys to send probes I think is likely to result that the client to server handshake flight will often contain 10-12k of data even if there are no 0-RTT application data. That might be a good investment, but I are we certain of the DoS properties of this? 

So my though was simply to be a bit conservative here and ensure that both endpoints know there are someone there that will report delivery of the probes before sending them. But, this is clearly something the QUIC WG can discuss and change.

> the endpoint has established a 1-RTT key.

Reviewer 2: @gorryfair replies (below, repeated to keep the thread intact):

> The reason why we didn't suggest this within DPLPMTUD, was two things:
>
> * Using the limited for IW for packet probes rapidly consumes the current IW space in bytes, and we were unconvinced of the importance to tenatively try a large PMTU during the first RTT.
> 
> * I was not aware of measurement data about how changing the initial flight of data would impact congestion collapse, since at this point there is no feedback from the receiver, and it seems, at least to me, to raise questions regarding stability with large numbers of flows and also as you note, the potential DoS opportunities.
>
> So, I'd be interested in what people think about this, and see any analysis and experience.

> the endpoint has established a 1-RTT key.

Reviewer 2: I think that the intent of the text is fine, I was only concerned about the execution, purely from a logical standpoint.  That is, `(handshake complete) && (got 1-RTT keys) == (handshake complete)` because `(got 1-RTT keys)` is a precondition for `(handshake complete)`.

As you both say, there are concerns with probing before this point.  Maybe you can, and indeed it is probably beneficial.  But maybe it is also risky, and we don't need that.  If anyone does want to probe before this point, they are off the books for now.  If anyone wants to explicitly allow that in the draft, that's a design change that needs support, justification, and a separate issue to track.  I don't think we can do that as part of this change.

> the endpoint has established a 1-RTT key.

Author: This seems like the best direction to me - I leave to you whether (got 1-RTT keys) needs to be mentioned as not, and included that in your PR in case you wish to edit (or not).

###  Sending QUIC DPLPMTUD Probe Packets

> DPLPMTU Probe packets consists of a QUIC Header and a payload containing a
> PING Frame and multiple PADDING Frames, this can implement
> "Probing using padding data" (see section 4.1 of {{!DPLPMTUD}}. 

Reviewer 2: This is too narrowly restrictive. I would instead say:

```suggestion
DPLPMTU probe packets are ack-eliciting packets.  Probe packets that use the
PADDING frame therefore implement "Probing using padding data", as defined in
Section 4.1 of {{!DPLPMTUD}}.
```

Author: Looks OK, the text needs at least to 
(1) describe how to construct a probe packet
(2) explain if the probe packet can carry data or only padding.

> These can be generated without affecting the transfer of other QUIC frames.
> The PING Frame is used to trigger generation of an acknowledgement.
> Multiple PADDING Frames are used together to control the length of the probe packet.  

Reviewer 2: Again, too prescriptive.

```suggestion
```

Author: Would you be happy replacing this with /A PING Frame can be used to trigger generation of an acknowledgement to the packet probe./

Reviewer 2: That's already implied by "ack-eliciting".

> These frames might not be retransmitted if a probe packet containing
> them is lost.  The frames consume congestion window,
> which could delay the transmission of subsequent application data.

Reviewer 2: This is incorrect.  PADDING and PING are not retransmitted on loss.  The fact that they are ack-eliciting causes them to consume congestion window, but this is less of a problem if they include other data.  Which they may well do (albeit with a higher risk of that data needing retransmission).

```suggestion

DPLPMTUD probe packets consume congestion window, so they could delay
the transmission of subsequent application.
```

Author: The initial sentence was already there - I thought it added nothing, so happy to see that removed. 
The fix to the second sentence is broken, but I think this might be /could delay subsequent transmission by the application/?

Reviewer 2: Updated the broken text.

###  Validating the QUIC Path with DPLPMTUD

QUIC provides an acknowledged PL, therefore a sender does not
implement the DPLPMTUD CONFIRMATION_TIMER while in the SEARCH_COMPLETE state.

> ###  Handling of ICMP PTB Messages by DPLPMTUD

Reviewer 2: ```suggestion
###  Handling of ICMP Messages by DPLPMTUD
```

An endpoint using DPLPMTUD requires the validation of any received PTB message 
before using the PTB information, as defined in section 4.6 of {{!DPLPMTUD}}.
In addition to UDP Port
validation, QUIC validates an ICMP message by using other PL
information (e.g., validation of connection identifiers (CIDs) in the
quoted packet of any received ICMP message). 

> The further considerations for processing ICMP messages described in the previous section also

Reviewer 2: ```suggestion

The considerations for processing ICMP messages described in {{icmp-pmtud}} also
```

over time.  The largest UDP payload an endpoint sends at any given time is
referred to as the endpoint's maximum packet size.

An endpoint SHOULD use Datagram Packetization Layer PMTU Discovery
({{!DPLPMTUD=I-D.ietf-tsvwg-datagram-plpmtud}}) or implement Path MTU Discovery
(PMTUD) {{!RFC1191}} {{!RFC8201}} to determine whether the path to a destination
will support a desired message size without fragmentation.
In the absence of these mechanisms, QUIC endpoints SHOULD NOT send IP packets
larger than the minimum QUIC packet size. 

All QUIC
packets other than PMTUD/DPLPMTUD probe packets SHOULD be sized to fit within the
maximum packet size to avoid the packet being fragmented or dropped
{{?RFC8085}}.

If a QUIC endpoint determines that the PMTU between any pair of local and remote

> IP addresses has fallen below the smallest allowed maximum packet size, it MUST immediately 

Reviewer 2: I think that Gorry's changes (recognizing the minimum on MPS in DPLPMTUD) are OK, but I would add "of 1200 bytes" for clarity.
